new/sdk/networkservicemesh/common/heal/client.go

```diff
@@ -0,0 +1,145 @@
+package heal
+
+import (
+	"context"
+	"runtime"
+	"time"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connection"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/networkservice"
+	"github.com/networkservicemesh/networkservicemesh/new/sdk/networkservicemesh/core/next"
+	"github.com/networkservicemesh/networkservicemesh/new/sdk/tools/extended_context"
+	"github.com/networkservicemesh/networkservicemesh/new/sdk/tools/serialize"
+)
+
+type healClient struct {
+	requestors        map[string]func()
+	closers           map[string]func()
+	reported          map[string]*connection.Connection
+	onHeal            []networkservice.NetworkServiceClient
+	client            connection.MonitorConnectionClient
+	eventReceiver     connection.MonitorConnection_MonitorConnectionsClient
+	updateExecutor    serialize.Executor
+	recvEventExecutor serialize.Executor
+	cancelFunc        context.CancelFunc
+}
+
+func NewClient(client connection.MonitorConnectionClient, onHeal ...networkservice.NetworkServiceClient) networkservice.NetworkServiceClient {
+	rv := &healClient{
+		onHeal:            onHeal,
+		requestors:        make(map[string]func()),
+		closers:           make(map[string]func()),
+		reported:          make(map[string]*connection.Connection),
+		client:            client,
+		updateExecutor:    serialize.NewExecutor(),
+		eventReceiver:     nil, // This is intentionally nil
+		recvEventExecutor: nil, // This is intentionally nil
+	}
+	rv.onHeal = append(rv.onHeal, rv)
```

Comment:
And it is not clear why here we add ourselves to `onHeal` array? I think  `healClient` already a piece of passed chain?

Reply:
You are correct in the case where the client is part of a pass through Endpoint.  To handle the case where the client is the origin of the chain, if a nil onHeal is passed in we add the Client itself as onHeal.